### PR TITLE
feat: RL recipe for ttsql

### PR DESCRIPTION
**Pull request by ShubyM**
We have an RL and SFT + RL recipe for gemma4 e2b. See a write up of how to run this [here](https://github.com/ShubyM/open-rl/blob/5d92abea0e425b0cfa0825c380ab2b78e08218b4/examples/rl/text-to-sql/rl-recipe.md). The commits also can be viewed individually to make the review easier

### CODE REVIEW

File: examples/rl/text-to-sql/README.md
Comment (chuangw6) [inline examples/rl/text-to-sql/README.md:1]: `README.md` and `rl-recipe.md` share a lot of overlapping context. Let's delete the current `README.md` and rename `rl-recipe.md` to `README.md` to keep everything in one single source of truth.

**Comment by chuangw6** [comment]
Thanks for the PR.

/hold 

reviewing...

File: examples/rl/text-to-sql/utils/plot.py
Comment (chuangw6) [inline examples/rl/text-to-sql/utils/plot.py:1]: To keep the example directory focused purely on the core training script (`texttosql_sft_grpo.py`), let's organize the rest:
- Move the supporting logic (plot_ttsql_curves.py, texttosql_grpo_utils.py, texttosql_rewards.py) into a subfolder e.g. `utils/`.
- Move the demonstration PNGs into a subfolder e.g. `results/`.

File: examples/rl/text-to-sql/texttosql_rewards.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  score += 0.25 * (min(len(predicted_rows), len(target_rows)) / max(len(predicted_rows), len(target_rows)))
+  # Value overlap across all cells (50%)
+  pred_vals = {repr(v) for row in predicted_rows for v in row}
+  target_vals = {repr(v) for row in target_rows for v in row}
+  if target_vals:
+    score += 0.5 * min(len(pred_vals & target_vals) / len(target_vals), 1.0)
+  return score
+
+
+def compute_sql_reward(
```
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_rewards.py]: nit: For the helper files, it’s tough to parse which functions are meant to be consumed by the main training loop versus internal math. Consider using `__all__ = ["compute_sql_reward"]` to explicitly mark public exports.

File: examples/rl/text-to-sql/rl-recipe.md
Comment (chuangw6) [inline examples/rl/text-to-sql/rl-recipe.md:1]: Great guide overall! To make the instructions more guided and user-friendly, I’d suggest restructuring the `rl-recipe.md` file into the following clean outline:

1. **Overview:** 
   * Elevator pitch (what the recipe trains & the end goal).
   * What the underlying script actually does.
   * Explanation of the core methods (SFT + RL).
   * Summary of basic configurations used. (can link to code)
2. **Prerequisites:** 
   * Machine hardware requirements.
   * Repository setup sequence (Clone repo, install Astral `uv`, apply the temporary vLLM Gemma 4 patch).
3. **Start the Server:** 
   * Clear breakdown for spinning up the serving infrastructure (Terminal 1: Sampler, Terminal 2: Gateway).
4. **Run the Training:** 
   * Sequential instructions for executing Full vs. RL-only phases.
5. **Result and graphs:** 
   * Baseline metrics expectations (+% execution matches).
6. **Chat with Model:** 
   * Provide an execution sandbox block using `tinker.ServiceClient` to test new queries dynamically.
7. **Troubleshooting**
   *  Add a common pitfalls section

Comment (chuangw6) [reply examples/rl/text-to-sql/rl-recipe.md:1]: Example draft generated by AI https://gist.github.com/chuangw6/d1438f5991aebfb01edb8f71161b51df.

File: examples/rl/text-to-sql/rl-recipe.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+The metrics and recipe log for this run are in:
+
+```text
+examples/rl/text-to-sql/artifacts/texttosql_sft_grpo_gemma4_e2b_rl_recipe_sft5_train100_vllm0191_no_bos/
+```
+
+## Step-by-Step Guide to Running the Recipe
+
+This guide assumes you have already cloned the repository. It is optimized for a machine with at least two GPUs, allowing you to run the sampler and gateway on separate devices.
```
Comment (chuangw6) [inline examples/rl/text-to-sql/rl-recipe.md]: Suggest making this prerequisite explicit, what minimum GPU model is required to run and mention GCE as an example and point it to the GCP doc e.g. https://docs.cloud.google.com/compute/docs/gpus/create-gpu-vm-a3u-a4

File: examples/rl/text-to-sql/rl-recipe.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+```bash
+curl -LsSf https://astral.sh/uv/install.sh | sh
+```
+
+Restart your shell or source your profile to make `uv` available in your PATH.
+
+### Step 2: Patch vLLM for Gemma 4
+
+Run the patch script to avoid duplicate module issues with Gemma 4 in vLLM. The script will automatically check if the patch is needed and apply it, or report if it is already in place.
```
Comment (chuangw6) [inline examples/rl/text-to-sql/rl-recipe.md]: Mark this as temporary step if so and mention the reason/context (github issue/PR) briefly.

Something like `Run the local patch script to resolve duplicate LoRA mapping constraints unique to Gemma 4:`

File: examples/rl/text-to-sql/rl-recipe.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+```text
+baseline eval @ step 0:
+execution_match=8%, exact_match=4%, similarity=0.372
+
+SFT eval @ step 5:
+execution_match=12%, exact_match=6%, similarity=0.494
+```
+
+### Example Before/After SFT
+
```
Comment (chuangw6) [inline examples/rl/text-to-sql/rl-recipe.md]: maybe we can let users to try the base model and trained model themselves by using our CLI? https://github.com/gke-labs/open-rl/tree/main/dev/tools#2-chat-with-an-adapter

File: examples/rl/text-to-sql/texttosql_gemma4_plain_notebook.ipynb
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_gemma4_plain_notebook.ipynb:1]: Please make sure the notebook can open properly. It currently renders as invalid/corrupted on my end. 
See https://github.com/ShubyM/open-rl/blob/feat/grpo-ttsql/examples/rl/text-to-sql/texttosql_gemma4_plain_notebook.ipynb

File: examples/rl/text-to-sql/texttosql_sft_grpo.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+  match: float = 1.0
+  error_penalty: float = -0.5
+  similarity: float = 0.0
+
+
+@chz.chz
+class Config:
+  model: ModelConfig
+  phase: str = "full"  # "full" | "sft_only" | "rl_only"
+  base_url: str = os.getenv("TINKER_BASE_URL") or os.getenv("OPEN_RL_BASE_URL") or BASE_URL
```
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_sft_grpo.py]: didn't we consolidate those envs already?

File: examples/sft/text-to-sql/texttosql_sft.py [hunk 6]
```diff
@@ -355,13 +353,13 @@ def build_examples(tokenizer, prompt_format: str, dataset_split, limit, require_
   return examples
 
 
-async def evaluate(sampler, tokenizer, alias, examples, config):
-  execution_match, similarity = 0.0, 0.0
+async def evaluate_metrics(sampler, tokenizer, alias, examples, max_tokens: int, seed: int) -> dict[str, float]:
```
Comment (chuangw6) [inline examples/sft/text-to-sql/texttosql_sft.py:356]: Right now, `texttosql_sft_grpo.py` imports helpers directly from the executable `texttosql_sft.py`.

Can we extract these into a shared file e.g. `examples/sft/text-to-sql/data_utils.py`
- DB Operations: run_sql(), sql_results_match(), clean_sql_for_execution(), normalize_sql().
- Token/Data mapping: build_examples(), render_training_texts(), make_datum().

This way, we can strip texttosql_sft.py down to pure config definitions and execution paths, and change texttosql_sft_grpo.py imports to fetch from `data_utils` instead of the executable training script.

File: examples/rl/text-to-sql/texttosql_sft_grpo.py
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_sft_grpo.py:1]: Suggest focusing the training script on a clean readable loop of data fetching, rollout logic, and standard backward passes, and offloading db/data/token/viz methods to somewhere else.

Same for `examples/sft/text-to-sql/texttosql_sft.py`

File: examples/rl/text-to-sql/texttosql_sft_grpo.py [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+    return f"{before[name] * 100:.1f}%->{after_sft[name] * 100:.1f}%->{after_rl[name] * 100:.1f}%"
+
+
+async def snapshot_eval(trainer: tinker.TrainingClient, alias: str, eval_examples: list[dict[str, Any]]) -> EvalMetrics:
+  """Snapshot current weights to a sampler and run the eval loop against it."""
+  sampler = await trainer.save_weights_and_get_sampling_client_async(name=alias)
+  return await evaluate_metrics(sampler, tokenizer, alias, eval_examples, max_tokens=config.dataset.eval_max_tokens, seed=config.seed)
+
+
+def make_rl_datum(rollout: dict[str, Any], advantage: float) -> types.Datum:
```
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_sft_grpo.py]: Offload to data_utils.py?

File: examples/rl/text-to-sql/texttosql_sft_grpo.py [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+    "after_rl_similarity": after_rl_metrics["similarity"],
+    "metrics_path": str(metrics_path),
+  }
+  result["final_state_path"] = final_state_path
+  if sft_state_path:
+    result["sft_state_path"] = sft_state_path
+  return result
+
+
+# *** Helpers used by the training phases above ***
```
Comment (chuangw6) [inline examples/rl/text-to-sql/texttosql_sft_grpo.py]: offload these helpers to texttosql_grpo_utils.py?

**Comment by droot** [comment]
Thanks @ShubyM for getting this in. Those graphs makes me super happy :)

I will defer to @chuangw6 to approve this.

File: examples/rl/text-to-sql/rl-recipe.md
Comment (ShubyM) [reply examples/rl/text-to-sql/rl-recipe.md:1]: I changed it but still needs some work to get to flow nicely I believe we can do this in a follow up

File: examples/rl/text-to-sql/rl-recipe.md (hunk 2 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/rl-recipe.md]: Done.

File: examples/rl/text-to-sql/texttosql_sft_grpo.py (hunk 5 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/texttosql_sft_grpo.py]: Fixed

File: examples/sft/text-to-sql/texttosql_sft.py (hunk 6 above)
Comment (ShubyM) [reply examples/sft/text-to-sql/texttosql_sft.py:356]: Seperated more of the functionality into a utils folder with 3 files helpers plot and rewards

File: examples/rl/text-to-sql/texttosql_sft_grpo.py (hunk 7 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/texttosql_sft_grpo.py]: understanding the concept of a datum i feel is fairly important to the actual loop so i left this in

File: examples/rl/text-to-sql/rl-recipe.md (hunk 3 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/rl-recipe.md]: Done.

File: examples/rl/text-to-sql/rl-recipe.md (hunk 4 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/rl-recipe.md]: Done.

File: examples/rl/text-to-sql/texttosql_sft_grpo.py
Comment (ShubyM) [reply examples/rl/text-to-sql/texttosql_sft_grpo.py:1]: I think I agree with the viz and db stuff but I think data and token being in the same file is fine as I think for someone using it end to end its easier to see how to start and how to end

File: examples/rl/text-to-sql/utils/plot.py
Comment (ShubyM) [reply examples/rl/text-to-sql/utils/plot.py:1]: Done.

File: examples/rl/text-to-sql/texttosql_rewards.py (hunk 1 above)
Comment (ShubyM) [reply examples/rl/text-to-sql/texttosql_rewards.py]: I think this is overkill for this file. I think people will read the training loop first and will see that compute_sql_reward is what it consumes and we are not importing via a wildcard

**Comment by ShubyM** [comment]
Anyone ever see the Github admin job fail?

**Comment by chuangw6** [comment]
> Anyone ever see the Github admin job fail?

just re-triggered the run and it's successful now
